093-split-metrics-reporter-into-modules.md

```diff
@@ -0,0 +1,51 @@
+# Split metrics-reporter into client and server modules
+
+## Current situation
+
+The metrics-reporter is currently a single artifact (JAR) that can be used with both Apache Kafka clients and servers. 
+
+## Motivation
+
+Having a single artifact is simple and works well for the basic functionality. We're now trying to add more advanced features and it's proved hard to keep the same logic for both client and server.
+
+For example [support for dynamic configurations](https://github.com/strimzi/metrics-reporter/issues/55) is a feature only available for servers. Adding support for this feature is difficult without impacting the client logic. I opened a PoC [PR](https://github.com/strimzi/metrics-reporter/pull/64), and while I got it to work, this caused a lot of complexity due to `KafkaPrometheusMetricsReporter` needing to support both clients and servers.
+
+Another benefit of having separate modules is the ability to enforce dependencies. Depending on where the reporter runs, a Kafka client or a Kafka server, different dependencies are available at runtime. We need to make sure that logic used by clients does not load any classes only available on servers. We hit this issue in the past ([issue #48](https://github.com/strimzi/metrics-reporter/issues/48)). This is not something that can easily be tested so currently it relies on maintainers being diligent when making code changes to ensure we don't hit this issue again in the future.
+
+I'm also considering adding support for [KIP-714](https://cwiki.apache.org/confluence/display/KAFKA/KIP-714%3A+Client+metrics+and+observability) to metrics-reporter. Again this feature is only implemented by server-side metrics reporters. Adding support for this feature will be addressed in a separate proposal.
+
+## Proposal
+
+I propose splitting the project into 2 Java modules:
+
+1. `client-metrics-reporter`: This module will only depend on kafka-clients and the Prometheus libraries. It will be used by Apache Kafka clients (Producer, Admin, Consumer, Connect and Streams) by setting the `metric.reporters` configuration to `io.strimzi.kafka.metrics.ClientMetricsReporter`. All the existing metrics-reporter configurations will stay the same. The differences are:
```

Comment:
Why not `KafkaClientMetricsReporter`, `KafkaServerMetricsReporter` and `YammerServerMetricsReporter`? It just reads better.

Reply:
Maybe we can add `prometheus` to the package name? That would keep  E.g. `io.strimzi.kafka.metrics.prometheus.ClientMetricsReporter`. (I suspect we might need other technologies as well one (maybe not so distant) day :-/)

Reply:
> we might need other technologies as well

In that case, I would expect a config parameter to configure the output format with a default value set to Prometheus for backwards compatibility.

Reply:
If we are to add extra reporters in the future, I expect users would pick the format by using different classes. For example if you use `io.strimzi.kafka.metrics.prometheus.ClientMetricsReporter` the export format is Prometheus. So I wouldn't add a new configuration to the reporter itself.

Reply:
Ok, makes sense. Didn't know Prom had push-based model.

Reply:
I don't think it is Prometheus itself - it would be more in the OpenTelemetry realm. But I think the push-based metrics are becoming more common.

Reply:
AFAIK, the is Prometheus Agent which is able to use the push-based model so writing metrics remotely as well.
Regarding the naming I am fine with `ClientMetricsReporter` but maybe within a `prometheus` package as suggested. It would be clearly Kafka even because the FQDM has `kafka` in the package as well.

Reply:
I like the suggestion of having a `prometheus` package. If we have consensus on that point, I'll make the change.

Reply:
I've pushed an update changing the packages to `io.strimzi.kafka.metrics.prometheus`.

Reply:
I know this could be a breaking change compared to the 0.1.0 but I think it's better doing it now.
+1 from me to change the modules and artifacts names to include prometheues naming as well. @tombentley @scholzj @Frawless @fvaleri @im-konge are you ok with that? (tagging you because already approved the proposal).

Reply:
You should probably clarify what exactly are you talking about.

Reply:
I assume that @mimaison metns that instead of having `io.strimzi.metrics-reporter-0.1.0.jar` as today, we would have `io.strimzi.prometheus.metrics-reporter-0.2.0.jar` instead. In the future, a different implementation would have a separate artifact with a different name (replacing "prometheus" with something else). This way if you want to use just the prometheus one you can get only the specific dependency and not bringing with you stuff which is unrelated but belongs to a different implementation.

Reply:
In any case if you decide to include Prometheus in the module name, it should absolutely not be in the group ID but only in the artifact ID.

Reply:
I guess that nobody knows how many dependencies would be useful or useless in different implementations. Also in what point in the future should we add another implementation? Having the name in the artifacts and modules now would look like there is another JAR containing different implementation (even if there is just Prometheus one).

I'm fine with adding `prometheus` package. For changing the names of the module and artifact - it could be changed in the future when another implementations will be added and the split will be needed, right?